Improve binary handling, add capability to prefer cache, add capability to cache on put.

Climax777 opened the pull request:
- Added missing dependencies to package.json
- Changed behaviour on put -> now caches on put as well
- Added all knox config options to allow s3rver for example
- Tested locally with s3rver

Climax777 committed:
Couple of changes:

- Added missing dependencies to package.json
- Changed behaviour on put -> now caches on put as well
- Added all knox config options to allow s3rver for example
- Tested locally with s3rver

Climax777 committed:
Missed secure option for knox

andrewjstone commented:
This patch looks good. I haven't used this code in years however. Are the knox config options required, and what happens when they aren't set? If they are required can you please add them to the example int he README.

Additionally, the README needs to be updated to reflect the location of the testing file changing. A reference to testing locally with s3erver would be good as well.

Thanks,
Andrew

Climax777 commented:
Hi

Sure I'll add that. The change in location isn't required, I'll revert
that.

The settings are required for s3-like servers other than AWS, like s3rver.

I'll add some instructions to the Readme for using s3rver.

Thanks for a great project, it's the only one of its kind I could find.

On Thu, 30 Mar 2017, 22:33 Andrew J. Stone <notifications@github.com> wrote:

> This patch looks good. I haven't used this code in years however. Are the
> knox config options required, and what happens when they aren't set? If
> they are required can you please add them to the example int he README.
>
> Additionally, the README needs to be updated to reflect the location of
> the testing file changing. A reference to testing locally with s3erver
> would be good as well.
>
> Thanks,
> Andrew
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-290536429>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAu7ncBu72ilmxdaz-7lZn23vZ9ccgglks5rrBGygaJpZM4Mubl0>
> .
>

Climax777 committed:
Changes as requested

- Reverted test file path change
- Added all the knox options
- Updated README

Climax777 commented:
@andrewjstone I've made the changes as requested, please have a look and let me know if you want any other changes.

Climax777 committed:
Prefer cache

This skips the unnecessary 304 costing, when you know that data doesn't get modified on S3

Climax777 committed:
Completely removed date concept

Again assuming that this interface is always used. No external S3 modifications are done

Climax777 committed:
Use putBuffer rather

Climax777 committed:
Revert putBuffer change

Climax777 committed:
Correctly handle binary data

Climax777 committed:
Still more binary handling

andrewjstone commented:
I don't think the cache related changes are correct. They only work in the single cache case where all changes go through s3asy. There is never any cache eviction, which allows permanent staleness if the documents are updated externally to s3asy or on another system with a different redis cache. The way the code was written the contents will always be consistent, and the cache is only used to prevent downloading of large files. 

This change makes it so the cache is used always, which while more efficient, is incorrect for the reasons stated above. It's a change to the semantics of how things work currently and will break other users. If the you really want to always use the cache first and don't care about the consistency of the files (which is probably fine since you are caching), I would suggest instead using a ttl on the redis keys. Since I think the original behavior is also useful, I suggest checking for the 'If-Modified-Since' header and if it is not set then using the new behavior, otherwise use the old behavior. Note that this also requires a change to the docs since it says the 'If-Modified-Since' header is always required.

The change to use binary seems to work, and I'm fine with it. I suppose it can be fixed later if it's an issue.

Finally, the last test currently fails due to the changes in cache control. If you change the behavior as I suggested please add a test that works with both modes of operation. Additionally, it's fine to always use a ttl with the suggested changes, as I can't see any other way of making it work correctly with both options.

```
 s3asy
    ✓ can put a file (495ms)
    ✓ can retrieve the same file
    ✓ has the file cached
    ✓ can retrieve the cached file
    ✓ does not have the file cached after 2 seconds (2007ms)
    ✓ can save and retrieve the original content again (103ms)
    1) can successfully bust the cache


  6 passing (3s)
  1 failing

  1) s3asy can successfully bust the cache:
     Uncaught AssertionError: undefined == true
      at test.js:89:18
      at Command.callback (/Users/andrewstone/s3asy/node_modules/cacheit/cache.js:76:14)
      at normal_reply (/Users/andrewstone/s3asy/node_modules/redis/index.js:721:21)
      at RedisClient.return_reply (/Users/andrewstone/s3asy/node_modules/redis/index.js:819:9)
      at JavascriptRedisParser.Parser.returnReply (/Users/andrewstone/s3asy/node_modules/redis/index.js:192:18)
      at JavascriptRedisParser.execute (/Users/andrewstone/s3asy/node_modules/redis-parser/lib/parser.js:562:12)
      at Socket.<anonymous> (/Users/andrewstone/s3asy/node_modules/redis/index.js:274:27)
      at readableAddChunk (_stream_readable.js:146:16)
      at Socket.Readable.push (_stream_readable.js:110:10)
      at TCP.onread (net.js:523:20)
```

Climax777 commented:
Hi Andrew, I should have done this on a different branch in retrospect.

What if I add an option to prefer redis? If set it uses the code path like
in my latest patches, otherwise it uses s3 first with the header instead?
Adding tests for this goes without saying.

I specifically require reduction of s3 costs and even with the header we
get charged a GET request.

On Fri, 31 Mar 2017, 18:00 Andrew J. Stone <notifications@github.com> wrote:

> I don't think the cache related changes are correct. They only work in the
> single cache case where all changes go through s3asy. There is never any
> cache eviction, which allows permanent staleness if the documents are
> updated externally to s3asy or on another system with a different redis
> cache. The way the code was written the contents will always be consistent,
> and the cache is only used to prevent downloading of large files.
>
> This change makes it so the cache is used always, which while more
> efficient, is incorrect for the reasons stated above. It's a change to the
> semantics of how things work currently and will break other users. If the
> you really want to always use the cache first and don't care about the
> consistency of the files (which is probably fine since you are caching), I
> would suggest instead using a ttl on the redis keys. Since I think the
> original behavior is also useful, I suggest checking for the
> 'If-Modified-Since' header and if it is not set then using the new
> behavior, otherwise use the old behavior. Note that this also requires a
> change to the docs since it says the 'If-Modified-Since' header is always
> required.
>
> The change to use binary seems to work, and I'm fine with it. I suppose it
> can be fixed later if it's an issue.
>
> Finally, the last test currently fails due to the changes in cache
> control. If you change the behavior as I suggested please add a test that
> works with both modes of operation. Additionally, it's fine to always use a
> ttl with the suggested changes, as I can't see any other way of making it
> work correctly with both options.
>
>  s3asy
>     ✓ can put a file (495ms)
>     ✓ can retrieve the same file
>     ✓ has the file cached
>     ✓ can retrieve the cached file
>     ✓ does not have the file cached after 2 seconds (2007ms)
>     ✓ can save and retrieve the original content again (103ms)
>     1) can successfully bust the cache
>
>
>   6 passing (3s)
>   1 failing
>
>   1) s3asy can successfully bust the cache:
>      Uncaught AssertionError: undefined == true
>       at test.js:89:18
>       at Command.callback (/Users/andrewstone/s3asy/node_modules/cacheit/cache.js:76:14)
>       at normal_reply (/Users/andrewstone/s3asy/node_modules/redis/index.js:721:21)
>       at RedisClient.return_reply (/Users/andrewstone/s3asy/node_modules/redis/index.js:819:9)
>       at JavascriptRedisParser.Parser.returnReply (/Users/andrewstone/s3asy/node_modules/redis/index.js:192:18)
>       at JavascriptRedisParser.execute (/Users/andrewstone/s3asy/node_modules/redis-parser/lib/parser.js:562:12)
>       at Socket.<anonymous> (/Users/andrewstone/s3asy/node_modules/redis/index.js:274:27)
>       at readableAddChunk (_stream_readable.js:146:16)
>       at Socket.Readable.push (_stream_readable.js:110:10)
>       at TCP.onread (net.js:523:20)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-290753623>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAu7nRJ8f0f3_aHAsWRsnMqXcg0PrgDqks5rrSMCgaJpZM4Mubl0>
> .
>

andrewjstone commented:
The option to prefer redis works for me as well :) I totally understand the need to avoid S3 as much as possible. Maybe also add a ttl option so your feature could be used with ttl if wanted.

Climax777 commented:
Great, I'm glad we have an agreement. Should I also set the ttl option for
the s3 preferred code path? I assume I set the cache's ttl?

On Fri, 31 Mar 2017, 18:14 Andrew J. Stone <notifications@github.com> wrote:

> The option to prefer redis works for me as well :) I totally understand
> the need to avoid S3 as much as possible. Maybe also add a ttl option so
> your feature could be used with ttl if wanted.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-290757750>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAu7ncSGIuaOL5tF2lb30w6419vlo4x1ks5rrSZygaJpZM4Mubl0>
> .
>

andrewjstone commented:
I don't think it's necessary to set the ttl in the S3 case since S3 will
tell us if we are stale in that case.

On Fri, Mar 31, 2017 at 12:29 PM, Pieter Jordaan <notifications@github.com>
wrote:

> Great, I'm glad we have an agreement. Should I also set the ttl option for
> the s3 preferred code path? I assume I set the cache's ttl?
>
> On Fri, 31 Mar 2017, 18:14 Andrew J. Stone <notifications@github.com>
> wrote:
>
> > The option to prefer redis works for me as well :) I totally understand
> > the need to avoid S3 as much as possible. Maybe also add a ttl option so
> > your feature could be used with ttl if wanted.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-290757750>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAu7ncSGIuaOL5tF2lb30w6419vlo4x1ks5rrSZygaJpZM4Mubl0>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-290761608>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcf-Udosp1L-OVaK5zWsP6VcbuZh8sJks5rrSnPgaJpZM4Mubl0>
> .
>

Climax777 commented:
Hi Andrew, inspecting `cacheit` I see the default TTL is 300s (5 minutes). Is this the desired behavior or should it default to 0 (infinite) for the S3-preferred solution?

andrewjstone commented:
Hmm, That is a good question. It's been so long since I've worked on this I
forgot about that. Great Catch. I believe the reason I added a timeout was
probably just to eliminate unnecessary caching/memory usage of infrequently
used items. Maybe just make the timeout larger, like an hour. Does that
make sense to you?

On Mon, Apr 3, 2017 at 1:14 AM, Pieter Jordaan <notifications@github.com>
wrote:

> Hi Andrew, inspecting cacheit I see the default TTL is 300s (5 minutes).
> Is this the desired behavior or should it default to 0 (infinite) for the
> S3-preferred solution?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-291050361>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcf-X_ZoiXDPe4zU_DbLL1S5tFJhlvlks5rsIBIgaJpZM4Mubl0>
> .
>

Climax777 commented:
@andrewjstone I just wanted to let you know I just haven't had the time yet, but I still plan on making the changes as discussed above.

andrewjstone commented:
No worries. Whenever you get around to it is fine. The project isn't going
anywhere :)

On Wed, May 3, 2017 at 8:14 AM, Pieter Jordaan <notifications@github.com>
wrote:

> @andrewjstone <https://github.com/andrewjstone> I just wanted to let you
> know I just haven't had the time yet, but I still plan on making the
> changes as discussed above.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andrewjstone/s3asy/pull/1#issuecomment-298893777>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcf-cDkdQmPV5fLYyAKiCvFozoE2vXmks5r2G-lgaJpZM4Mubl0>
> .
>

Climax777 committed:
Readded old behaviour where staleness is not acceptable

Climax777 committed:
Added flag for saving on put

Climax777 committed:
Improved tests, added doc entries

Climax777 commented:
@andrewjstone Almost a year later, and at last, I've addressed your concerns...

Please take a look and let me know if you have any suggestions for merging. Though the API has changed, old behaviour is maintained by default and shouldn't affect other users.
